flang/lib/Lower/Bridge.cpp

```diff
@@ -1566,9 +1566,7 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                 // Conversions should have been inserted by semantic analysis,
                 // but they can be incorrect between the rhs and lhs. Correct
                 // that here.
-                auto addr =
-                    fir::getBase(isPointer ? genExprValue(assign.lhs, stmtCtx)
-                                           : genExprAddr(assign.lhs, stmtCtx));
+                auto addr = fir::getBase(genExprAddr(assign.lhs, stmtCtx));
```

Comment:
I'm not sure I'm in favor of this change in lowering semantics. The value of a pointer seems like it should be the "addressable" value (a bare reference or descriptor). We have these cases (using C syntax for more clarity):

`p` is a pointer-to-T
- genExprAddr(`p`)   was: `&p`  change: `p`
- genExprValue(`p`)  was: `p`  change: `*p`
- deref `p` was: `*p`

Reply:
Pointer dereference are not explicit in Fortran syntax/evaluate::Expression. The fact we are or not dereferencing the pointer depends on the context where the expression appears. From my observations, the number of places where a pointer is not dereferenced is quite limited:
- Allocate/Deallocate statement.
- pointer assignment
- function calls, when passing to an allocatable/pointer dummy.
- inquiry intrinsics

That leaves a vast majority of places where expression that are pointer designators must be dereferenced. Typically, the 52 places that are currently using `genExprValue()`, but the one where this comment is at, do expect they will get `*p`, not an address from the expression. So if we go for the approach you are suggesting, we are pushing to most statement implementation the responsibility to check if they have pointer/allocatable for all expressions where they just want a simple value for a bounds/lengths/boolean. That seems both risky (most likely people won't think about this case), and an increase in code complexity.
Hence my personal opinion lies in having `genExprValue` returning the target address, and leave it to the few places that actually want the pointer address to make a specific call to do a `genMutableBoxValue`.

Reply:
The rub is whether lowering should take a syntactic view or a semantic view of these. From an IR sense, a syntactic view is ambiguous. From a front-end sense, a semantic view is risky.

A syntax-based take is ok here for now. Although, I'd rather have predictable semantics in loweing whenever possible.